### PR TITLE
Enable setting of graph partitioner used in cell partitioning from Python

### LINKED ISSUE #1859: Enable choice of ParMETIS, SCOTCH or KaHIP partitioners from Python
It is possible to pass the chosen partitioner to the mesh creation functions from C++, but this is not possible from Python. What is needed is to expose the different partitioning functions in Python.

### PR DESCRIPTION
**Pull request by garth-wells**
This was previously only possible from the C++ interface.

Fixes #1859.

### CODE REVIEW

File: python/dolfinx/wrappers/caster_mpi.h [hunk 1]
```diff
@@ -38,14 +38,23 @@ class type_caster<dolfinx_wrappers::MPICommWrapper>
   // Python to C++
   bool load(handle src, bool)
   {
-    // Simplified version of isinstance(src, mpi4py.MPI.Comm) - avoids
-    // segfault when pybind11 tries to convert some other random type to
-    // MPICommWrapper
-    if (not hasattr(src, "Allgather"))
+    // NOTE: This check would be safer than hasattr but does not evaluate to
+    // true.
+    // if (PyObject_TypeCheck(src.ptr(), &PyMPIComm_Type))
```
Comment (jhale) [inline python/dolfinx/wrappers/caster_mpi.h:43]: @chrisrichardson Any idea why this check doesn't work? Perhaps you've been here before.

File: python/test/unit/mesh/test_mesh_partitioners.py [hunk 2]
```diff
@@ -20,13 +20,37 @@
 assert (tempdir)
 
 
-@pytest.mark.parametrize("partitioner", [partition_cells_graph])
+partitioners = [dolfinx.cpp.graph.partitioner()]
```
Comment (jhale) [inline python/test/unit/mesh/test_mesh_partitioners.py:23]: Should be empty list? Existence of default partitioner covered in next try/except statement.

File: python/test/unit/mesh/test_mesh_partitioners.py [hunk 3]
```diff
@@ -20,13 +20,37 @@
...
+try:
+    from dolfinx.cpp.graph import partitioner
+    partitioners.append(partitioner())
+except ImportError:
+    raise
+try:
+    from dolfinx.cpp.graph import partitioner_scotch
+    partitioners.append(partitioner_scotch())
+except ImportError:
+    raise
```
Comment (jhale) [inline python/test/unit/mesh/test_mesh_partitioners.py:33]: Should be pass? We can build just with e.g. Parmetis or KaHiP now?

**Review by jhale — CHANGES_REQUESTED** [review]

File: python/test/unit/mesh/test_mesh_partitioners.py (hunk 3 above)
Comment (garth-wells) [reply python/test/unit/mesh/test_mesh_partitioners.py:33]: Not yet.

File: python/setup.py [hunk 4]
```diff
@@ -47,7 +47,8 @@ def build_extension(self, ext):
                       f'-DPython3_LIBRARIES={sysconfig.get_config_var("LIBDEST")}',
                       f'-DPython3_INCLUDE_DIRS={sysconfig.get_config_var("INCLUDEPY")}']
 
-        cfg = 'Debug' if self.debug else 'Release'
+        # cfg = 'Debug' if self.debug else 'Release'
```
Comment (jhale) [inline python/setup.py:50]: Remember to backout.

**Review by jhale — APPROVED** [review]

**Pull request merged**